### ISSUE TITLE
[API Proposal]: Use value semantics for System.Collections.Immutable types

### ISSUE THREAD

**Issue by kalleguld** [dotnet/runtime#59901]
### Background and motivation

With the new `record` types having value semantics, a new set of collections with the same semantics would be a welcome support.

`ImmutableList`, `ImmutableSet`, `ImmutableArray` and friends would make a good candidate for making a set of collection types with value semantics.
This would be done by overriding `GetHashCode()`, `Equals`, `operator==`and `operator!=`.

### API Proposal

```C#
namespace System.Collections.Immutable
{
    public class ImmutableList<T> 
    {

        public static bool operator==(ImmutableList<T> lhs, ImmutableList<T> rhs);
        public static bool operator!=(ImmutableList<T> lhs, ImmutableList<T> rhs);
        public override int GetHashCode();
    }
}
```     

### API Usage

```C#
class Program
{
    static void Main(string[] args)
    {
        var p1 = new PersonRecord(
            "Arthur",
            ImmutableList.Create("Conan", "Doyle")
        );
        var p2 = new PersonRecord(
            "Arthur",
            ImmutableList.Create("Conan", "Doyle")
        );
        // This should now return true:
        Console.WriteLine($"Are the two records the same?: {p1 == p2}");
    }
}

record PersonRecord(
    string Firstname,
    IEnumerable<string> LastNames
);
```     

### Risks

Programs relying on the affected types having reference semantics would fail, and would have to use `Object.ReferenceEquals(a,b)` instead.

**Comment by Clockwork-Muse** [comment]
> Programs relying on the affected types having reference semantics would fail, and would have to use `Object.ReferenceEquals(a,b)` instead.

Yeah, this is what's known as a "breaking change", and for something like this it's a big one; probably this would be disqualified.

Note that value semantics, even before the inclusion of `record` types, was possible to specify - it was a deliberate decision to not do value semantics for the immutable collections.

Note, too, that in almost all cases separate collections will have different contents, so value semantics aren't a particular benefit (as opposed to being explicit that you're checking two collections for identical contents).

**Comment by kalleguld** [comment]
> Note that value semantics, even before the inclusion of record types, was possible to specify - it was a deliberate decision to not do value semantics for the immutable collections.

Do you know the reasoning behind this decision? 
Are there any disadvantages to having value semantics on immutable types?
Are there other Immutable collection libraries with value semantics?

> Note, too, that in almost all cases separate collections will have different contents, so value semantics aren't a particular benefit (as opposed to being explicit that you're checking two collections for identical contents).

The same could be said for any class,so there should be no reason to make `record`s. The problem is that the value-semantics of records break down if they contain a collection (or if they contain any reference-semantic type).

**Comment by Clockwork-Muse** [comment]
> Do you know the reasoning behind this decision?

No, but I suspect it's mostly to mimic the other collection types.

> Are there any disadvantages to having value semantics on immutable types?

(Strictly speaking the additional checks maybe impose a slight performance penalty, even with early exits, but this is likely to be small, otherwise) No.

> Are there other Immutable collection libraries with value semantics?

Some languages make collection types in general have value equality (Rust).  Still, often collections aren't regarded as "value-ish" types, because...

> The same could be said for any class,so there should be no reason to make `record`s.

That's not what's conceptually going on with value-equal types  - they're being treated as a single element, as if they were a language primitive.  Viewed from this perspective, collections usually aren't "single things" to compare.

> The problem is that the value-semantics of records break down if they contain a collection (or if they contain any reference-semantic type). There's no easy way of comparing two records if they contain a list.

Easy?  Maybe not completely.  The assumption is that if your `record` contains non-value-semantic types that you manually implement `Equals`/`HashCode` (you can take a shortcut here by having all the value-semantic types in an internal record/struct that auto-derives those, then wrap that in the public type that compares all the other fields). 
Note that not even all record types will base their value equality on all their fields (the common case of database records using just their ID comes to mind).  Depending on your domain, you may have to worry about string equality as well (casing, cultural canonicalization rules...).  In the end, you're still responsible for some areas of domain specification.

**Comment by kalleguld** [comment]
> No, but I suspect it's mostly to mimic the other collection types.

But they are not like the other collection types though. They are Immutable. 

The reason you normally don't want to use value-semantics in collections is because you can change them. Not so with Immutable Collections.

> That's not what's conceptually going on with value-equal types - they're being treated as a single element, as if they were a language primitive.

I disagree. If you have two *immutable* lists where all their elements are equal, the lists themselves should be equal. There is no way of telling them apart, except for getting their memory address. 
If you have two immutable objects you can't tell apart, they should be equal.
We do the same for strings, records and tuples, why shouldn't there be a standard way for collections? And if not in System.Collections.Immutable, then where?

In other words, I think the distinguishing factor for value/reference semantics should be immutability, not "single-element"ness. After all, neither tuples nor strings nor structs are single-elemented. 

> The assumption is that if your record contains non-value-semantic types that you manually implement Equals/HashCode

I know, hence this proposal :) I like the ease of use of records, having to re-implement Equals/HashCode to circumvent (what I think is) a shortcoming of the standard library seems like the wrong way around things. I would rather have a list type with value semantics.

**Comment by Frassle** [comment]
> In other words, I think the distinguishing factor for value/reference semantics should be immutability, not "single-element"ness. After all, neither tuples nor strings nor structs are single-elemented.

Note that F# has it's own immutable collections (`FSharp.Core.List`, `FSharp.Core.Map`, `FSharp.Core.Set`) which do have value semantics. I've found that quite useful, but I could understand this proposal not being doable because it's a pretty significant breaking change to behaviour.

**Comment by eiriktsarpalis** [comment]
Regardless of any potential benefits structural equality might bring to immutable collections, adding it at this point would be a significant breaking change. I would recommend using a custom `IEqualityComparer<ImmutableList<T>>` implementation if structural equality is necessary in your application.
